tree/treeplayer/inc/ROOT/TVec.hxx

```diff
@@ -0,0 +1,381 @@
+}
+
+template <typename T, typename V>
+auto operator/(const TVec<T> &v, const V &c) -> TVec<decltype(v[0] / c)>
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) {return t/c;});
+}
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
+TVec<char> operator>(const TVec<T> &v, const V &c)
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) -> char {return t>c;});
+}
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
+TVec<char> operator>=(const TVec<T> &v, const V &c)
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) -> char {return t>=c;});
+}
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
+TVec<char> operator==(const TVec<T> &v, const V &c)
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) -> char {return t==c;});
+}
+
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
+TVec<char> operator!=(const TVec<T> &v, const V &c)
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) -> char {return t!=c;});
+}
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
+TVec<char> operator<=(const TVec<T> &v, const V &c)
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) -> char {return t<=c;});
+}
+
+template <typename T, typename V, typename D = typename std::enable_if<!ROOT::Detail::VecOps::TIsTVec<V>::value, int>::type>
+TVec<char> operator<(const TVec<T> &v, const V &c)
+{
+   return ROOT::Internal::VecOps::Operate(v, [&c](const T& t) -> char {return t<c;});
+}
+///@}
+
+/** @name Math Operators with TVecs
+ *  Math operators involving TVecs
+*/
+///@{
+
+template <typename T, typename V>
+auto operator+(const TVec<T> &v0, const TVec<V> &v1) -> TVec<decltype(v0[0] + v1[0])>
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "+", [](const T& t, const V &v) {return t+v;});
+}
+
+template <typename T, typename V>
+auto operator-(const TVec<T> &v0, const TVec<V> &v1) -> TVec<decltype(v0[0] - v1[0])>
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "-", [](const T& t, const V &v) {return t-v;});
+}
+
+template <typename T, typename V>
+auto operator*(const TVec<T> &v0, const TVec<V> &v1) -> TVec<decltype(v0[0] * v1[0])>
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "*", [](const T& t, const V &v) {return t*v;});
+}
+
+template <typename T, typename V>
+auto operator/(const TVec<T> &v0, const TVec<V> &v1) -> TVec<decltype(v0[0] / v1[0])>
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "/", [](const T& t, const V &v) {return t/v;});
+}
+
+template <typename T, typename V>
+TVec<char> operator>(const TVec<T> &v0, const TVec<V> &v1)
```

Comment:
I know `vector<bool>` is evil, but `vector<char>` looks very weird in a public signature.
maybe we can alias is to `TVec::Bool_t`, or maaaaaaaaybe just maybe we can actually use `vector<bool>`: what would break?

Reply:
This signature is a consequence of a more general design choice which consists in not using TVec<bool> but rather something which would not rely on the specialised implementation of the collection.
Let me tackle the questions backwards.
- The forthcoming buffer protocol implementation and all the migration of "contiguous blocks of memory" from ROOT to NumPy would fail, right?
- Wouldn't vectorisation performance be impacted by the usage of a vector<bool>?
- The usage of a type which seems to suggest it's a boolean but it's not might be more elegant but wouldn't it be even weirder than the signature with char?
Maybe a solution could really consist in encapsulating the present TVec implementation as TVecImpl in a new TVec template which could well be defined as TVec<bool> but would internally be implemented as a TVecImpl<char>.

Reply:
afaik buffer protocol and vectorisation could work with `vector<bool>` as it still stores data in contiguous memory. it might make it awkward though, if it requires special handling.

anyway I was just wondering, but no matter, let's stay away from `vector<bool>` :)
what about `vector<int>` or `vector<short>` though? they might look less surprising when people read the signature.

Reply:
I agree (see above), except it can't be auto-deduced because we don't want `vector<bool>` to pop up in bad places

Reply:
Indeed, I just checked and the comparison type is a `bool` in C++, not an `int`, but in hardware, using an integer type (either `int` or `long`, depending on the compared type) is probably best.